components/prism-odin.js

```diff
@@ -0,0 +1,61 @@
+(function (Prism) {
+	var escapes = /\\(?:["'\\abefnrtv]|0[0-7]{2}|U[\dA-Fa-f]{6}|u[\dA-Fa-f]{4}|x[\dA-Fa-f]{2})/.source;
+
+	function withEscapes(pattern, flags) {
+		return RegExp(pattern.replace(/<escapes>/g, escapes), flags);
+	}
+
+	Prism.languages.odin = {
+		/**
+		 * The current implementation of multiline comments doesn't support nesting.
+		 */
+		'comment': [
+			{
+				pattern: /\/\*[\s\S]*?(?:\*\/|$)/,
+				greedy: true
+			},
+			{
+				pattern: /#![^\n\r]*/,
+				greedy: true
+			},
+			{
+				pattern: /\/\/[^\n\r]*/,
+				greedy: true
+			}
+		],
+
+		/**
+		 * Should be found before strings because of '"'"- and '`'`-like sequences.
+		 */
+		'char': {
+			pattern: withEscapes(/'(?:<escapes>|[^\n\r'\\])'/.source),
+			inside: {
+				'symbol': withEscapes(/<escapes>/.source)
+			}
+		},
+
+		'string': [
+			{
+				pattern: /`[^`]*`/,
+				greedy: true
+			},
+			{
+				pattern: withEscapes(/"(?:<escapes>|[^\n\r"\\])*"/.source),
+				greedy: true,
+				inside: {
+					'symbol': withEscapes(/<escapes>/.source)
+				}
+			}
+		],
+
+		'number': /(?:\b0(?:b[01_]+|d[\d_]+|h_*(?:(?:(?:[\dA-Fa-f]_*){8}){1,2}|(?:[\dA-Fa-f]_*){4})|o[0-7_]+|x[\dA-F_a-f]+|z[\dAB_ab]+)\b|(?:\b\d+(?:\.\d*)?|\B\.\d+)(?:[Ee][+-]?\d*)?[ijk]?)/,
+
+		'boolean': /\b(?:_|false|nil|true)\b/,
+
+		'keyword': /\b(?:asm|auto_cast|bit_set|break|case|cast|context|continue|defer|distinct|do|dynamic|else|enum|fallthrough|for|foreign|if|import|in|map|matrix|not_in|or_else|or_return|package|proc|return|struct|switch|transmute|typeid|union|using|when|where)\b/,
+
+		'function': /\b\w+(?=\()/,
+
+		'operator': /\+\+|---?|->|\.\.[<=]?|(?:&~|[-!*+/=~]|[%&<>|]{1,2})=?|[#$(),.:;?@\[\]^{}]/
```

Comment:
We typically differentiate between operators and punctuation. 

Operators are e.g. addition, division, bitwise or, comparison, variable assignment, or anything else that creates/changes data. We are typically pretty broad with we allow being operators, so when in doubt, make it an operator.

Punctuation are symbols that glue together other pieces of syntax. E.g. separators, delimiters, and brackets for grouping.

---

Also, what is `#`?

```suggestion
		'operator': /\+\+|---?|->|\.\.[<=]?|(?:&~|[-!*+/=~]|[%&<>|]{1,2})=?|[#$?@^]/,
		'punctuation': /::|[(),.:;\[\]{}]/
```

Reply:
`#` starts a tag [compiler] directive. The [manual] is flooded with hashed, but you can still find `#align`, `#assert`, `#no_bounds_check`, etc. there.

Reply:
I personally think of operators as functions. In this sense, an operator converts values of its parameters (or operands) into a result. `[(),.:;\[\]{}]` do not produce any values, so they are punctuation marks.

If we do not take into account the [opinion](https://github.com/odin-lang/Odin/wiki/Keywords-and-Operators) of the author, then I will have to make more corrections, because

- `---` is not an operator, too. It’s an identifier for an undefined value. Can’t figure out how to classify it;

- `..`, `..<`, `..=` are punctuation marks, since they denote intervals just like a `:`;

- `->` is just a punctuation mark. It points to the type of the return value;

- `$` indicates a “constant parameter”, which makes the procedure reject non-constant arguments. This thing is a tricky one. `$` is not related to `const` in C. (In Odin you cannot assign a value to a parameter anyway. By the way, you can’t take its address either.) `$` doesn’t modify the value. It just forbids variables. Is `$` an operator or a punctuation sign?

- `@` indicates a decorator. Not an operator.

`?` is a type assertion operator. `&` is an operator, since it returns the address of the value. What do you think about `^` (pointer type)?

Reply:
> `#` starts a tag [compiler] directive. The [manual] is flooded with hashed, but you can still find `#align`, `#assert`, `#no_bounds_check`, etc. there.

In that case, we should probably have a token `'directive': { pattern: /#\w+/, alias: 'property' }`.

> * `---` is not an operator, too. It’s an identifier for an undefined value. Can’t figure out how to classify it;

This should probably be its own token. E.g. `'undefined': { pattern: /---/, alias: '<alias>' }` with an 'operator' or maybe even 'important' alias.

> * `..`, `..<`, `..=` are punctuation marks, since they denote intervals just like a `:`;

I would say that they create a new range, so they are operators.

Regarding `:`: They are used for slices and `:` is just a separator to separate the low and high bounds. So I think `:` should be punctuation.

> * `->` is just a punctuation mark. It points to the type of the return value;

Agreed.

> * `$` indicates a “constant parameter”, which makes the procedure reject non-constant arguments. This thing is a tricky one. `$` is not related to `const` in C. (In Odin you cannot assign a value to a parameter anyway. By the way, you can’t take its address either.) `$` doesn’t modify the value. It just forbids variables. Is `$` an operator or a punctuation sign?

`$` marks a "variable" as a compile-time constant, right? From what I see, they are pretty much the same as C++ templates.

Doesn't that make `$` hugely significant? Maybe we should even give it its own token with an `important` alias?

> * `@` indicates a decorator. Not an operator.

Good point, let's make it punctuation.

> `?` is a type assertion operator. `&` is an operator, since it returns the address of the value. What do you think about `^` (pointer type)?

I would say that `^` operates on a type to create a pointer type, so I would say that it's an operator.

Reply:
> `$` marks a "variable" as a compile-time constant, right?

Yes, only constants can be used as arguments.

> I would say that `^` operates on a type to create a pointer type, so I would say that it's an operator.

I completely agree with you.

> So I think `:` should be punctuation.

Yes.

> > `..`, `..<`, `..=`
> 
> I would say that they create a new range, so they are operators.

True! They create new values. How could I miss that?

Reply:
> $ marks a "variable" as a compile-time constant, right? From what I see, they are pretty much the same as C++ templates.

I found the following snippet in the Odin overview (and fixed quite a bit):

```odin
package main

import "core:mem"

my_new :: proc($T: typeid) -> ^T {
  return (^T)(mem.alloc(size_of(T), align_of(T)))
}

main :: proc() {
  ptr := my_new(int)
}
```

This confirms that we can use `$` to mark a variable, which holds a type name. So yes, in this case the operator is similar to `typename` in C++.

Reply:
Done. See babdd8c